parsec/backend/postgresql/pki.py

```diff
@@ -0,0 +1,439 @@
+        submitter_der_x509_certificate: bytes,
+        submit_payload_signature: bytes,
+        submit_payload: bytes,
+        submitted_on: DateTime,
+    ) -> None:
+        """
+        Raises:
+            PkiEnrollmentCertificateAlreadySubmittedError
+            PkiEnrollmentAlreadyEnrolledError
+        """
+        submitter_der_x509_certificate_sha1 = hashlib.sha1(submitter_der_x509_certificate).digest()
+        async with self.dbh.pool.acquire() as conn, conn.transaction():
+            # Assert enrollment_id not used
+            row = await conn.fetchrow(
+                *_q_get_pki_enrollment_for_update(
+                    organization_id=organization_id.str, enrollment_id=enrollment_id
+                )
+            )
+            if row:
+                raise PkiEnrollmentIdAlreadyUsedError()
+
+            # Try to retrieve the last attempt with this x509 certificate
+            rep = await conn.fetch(
+                *_q_get_pki_enrollment_from_certificate_sha1(
+                    organization_id=organization_id.str,
+                    submitter_der_x509_certificate_sha1=submitter_der_x509_certificate_sha1,
+                )
+            )
+            for row in rep:
+                enrollment_state = row["enrollment_state"]
+                if enrollment_state == PkiEnrollmentStatus.SUBMITTED.value:
+                    if force:
+                        await conn.execute(
+                            *_q_cancel_pki_enrollment(
+                                organization_id=organization_id.str,
+                                enrollment_id=row["enrollment_id"],
+                                enrollment_state=PkiEnrollmentStatus.CANCELLED.value,
+                                cancelled_on=submitted_on,
+                            )
+                        )
+                        await send_signal(
+                            conn,
+                            BackendEvent.PKI_ENROLLMENT_UPDATED,
+                            organization_id=organization_id,
+                        )
+
+                    else:
+                        raise PkiEnrollmentCertificateAlreadySubmittedError(
+                            submitted_on=row["submitted_on"]
+                        )
+                elif enrollment_state in [
+                    PkiEnrollmentStatus.REJECTED.value,
+                    PkiEnrollmentStatus.CANCELLED.value,
+                ]:
+                    # Previous attempt was unsuccessful, so we are clear to submit a new attempt !
+                    pass
+                elif enrollment_state == PkiEnrollmentStatus.ACCEPTED.value:
+                    # Previous attempt end successfully, we are not allowed to submit
+                    # unless the created user has been revoked
+                    assert row["accepted"] is not None and row["accepter"] is not None
+
+                    row = await conn.fetchrow(
+                        *_q_get_user_from_device_id(
```

Comment:
we should also have a FOR UPDATE here... I'm wondering if we shouldn't use a advisory lock instead to simplify this locking logic (so we would have one big lock for user/device creation per organization)